### PR TITLE
docs: Provide scaler governance definition

### PR DESCRIPTION
**Pull request by tomkerkhove**
Provide scaler governance definition

### Checklist

- [x] Commits are signed with Developer Certificate of Origin (DCO)

Relates to #5

### CODE REVIEW

**Comment by tomkerkhove** [comment]
Took inspiration on https://github.com/kedacore/governance/pull/18 by @jeffhollan

File: SCALERS.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+- A scaler MUST have e2e and unit tests to guarantee its quality
+- A scaler MUST be backwards compatible until the next KEDA major version
+- A scaler CAN be deprecated when:
+  - Maintainers believe it is best for the project, after a vote
+  - The scaler owner(s) are not responsive and support volume is too high
+  - The technology is proven to be no longer used/abandoned
+- Scaler MUST follow KEDA release cycles
+- A scaler MUST be supported
+
+### Choosing a home for external scalers
```
Comment (tomkerkhove) [inline SCALERS.md:45]: Something we can evaluate would be to introduce a `keda-contrib` org for external scalers to centralize them; but give external maintainers the required permissions.

Although I don't think we need that at this point in time.

File: SCALERS.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+The general rules of thumb for built-in scalers are the following:
+
+- They must live in the [core KEDA repository](http://github.com/kedacore/keda)
+- It follows the [KEDA core release cycle](https://github.com/kedacore/keda/blob/main/ROADMAP.md#upcoming-release-cycles)
+- It has no additional infrastructure requirements
+- It is general-purpose and not specific to a single end-user
+- It is written in Go
+
+If all of the above are true, then then it is a potential case for a new built-in scaler. We recommend opening a feature request to propose and discuss it before doing the actual implementation though. This is to avoid throwing things away if it does not meet the needs after all.
```
Comment (JorTurFer) [inline SCALERS.md:28]: Duplicated "then"

File: SCALERS.md (hunk 1 above)
Comment (JorTurFer) [reply SCALERS.md:45]: I like the idea of having another org more open but maybe we could just invite them to current org and create different teams inside.
Wdyt?

Comment (JorTurFer) [reply SCALERS.md:45]: I mean, using current org we "create a group feeling", people could be more engaged being part of kedacore.

**Review by JorTurFer — APPROVED** [review]
I agree with the proposal, my comments are not blockers

File: SCALERS.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+- A scaler MUST be written in Go
+- A scaler owner MUST be assigned (individual(s)/company) and will be tagged for bug reports
+- A scaler MUST have e2e and unit tests to guarantee its quality
+- A scaler MUST be backwards compatible until the next KEDA major version
+- A scaler CAN be deprecated when:
+  - Maintainers believe it is best for the project, after a vote
+  - The scaler owner(s) are not responsive and support volume is too high
+  - The technology is proven to be no longer used/abandoned
+- Scaler MUST follow KEDA release cycles
+- A scaler MUST be supported, as per [our support policy](SUPPORT.md), by the scaler owner
```
Comment (tomkerkhove) [inline SCALERS.md:43]: I need to clarify that we'll also use codeowners to automatically assign PRs to scaler owners

File: SCALERS.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
+# Scaler Governance
```
Comment (tomkerkhove) [inline SCALERS.md:1]: "scaler owners" -> "scaler maintainers"

File: SCALERS.md (hunk 1 above)
Comment (tomkerkhove) [reply SCALERS.md:45]: > I like the idea of having another org more open but maybe we could just invite them to current org and create different teams inside.

If we do that, then there is no longer a distinction between "official" external scalers and the ones maintained by the community as pointed out in the doc.

I'll resolve this and we can still introduce this later on unless @zroubalik has some thoughts on this.

**Review by ahmelsayed — APPROVED** [review]

**Comment by jeffhollan** [comment]
Looks good - my main comment is I don't think we necessarily need to require all scalars have a maintainer - but scalars that don't need to be clearly identified and users should consume at their own risk (maintainers won't promise to fix bugs or anything else, and will be marked as 'help needed')

**Comment by tomkerkhove** [comment]
> Looks good - my main comment is I don't think we necessarily need to require all scalars have a maintainer - but scalars that don't need to be clearly identified and users should consume at their own risk (maintainers won't promise to fix bugs or anything else, and will be marked as 'help needed')

Something like "Community maintained"? What would be the stimulation for anyone to become the maintainer then? Won't that end up with all scalers be community maintained?

File: SCALERS.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+If all of the above are true, then it is a potential case for a new built-in scaler. We recommend opening a feature request to propose and discuss it before doing the actual implementation though. This is to avoid throwing things away if it does not meet the needs after all.
+
+### Requirements for a built-in scaler
+
+In order to contribute a scaler to KEDA's core, it must meet the following requirements:
+
+- A scaler MUST be written in Go
+- A scaler maintainer(s) MUST be assigned (individual(s)/company) and will be tagged for bug reports
+- A scaler MUST have e2e and unit tests to guarantee its quality
+- A scaler MUST be backwards compatible until the next KEDA major version
```
Comment (zroubalik) [inline SCALERS.md:37]: I would omit this. Next Major version of KEDA is something, that I'd like to avoid in near future.
I think it is safe to deprecate some fields of a scaler, if we have some deprecation period, for example 2 releases, where we support both new and old functionality, where possible.

File: SCALERS.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  - The scaler maintainer(s) are not responsive and support volume is too high
+  - The technology is proven to be no longer used/abandoned
+- Scaler MUST follow KEDA release cycles
+- A scaler MUST be supported, as per [our support policy](SUPPORT.md), by the scaler maintainer(s) and will be automatically assigned to PRs. This will be done by using GitHub's Codeowners.
+
+### Choosing a home for external scalers
+
+Given the nature of external scalers, we recommend that they are maintained by the community and have a home outside of the KEDA organization. The reason for this is that it sets clear expectations for the community in terms of ownership and support.
+
+There have been exceptions where they have been added to the KEDA organization, but that is because they are governed and co-maintained by the community. Given that they are part of the KEDA organization, the maintainers are responsible for the quality of the scaler, timelines, etc.
```
Comment (zroubalik) [inline SCALERS.md:49]:
```
Given that they are part of the KEDA organization, the maintainers are responsible for the quality of the scaler, timelines, etc.
```

How should I understand this? KEDA core maintainers are responsible for these scalers? For example this scaler: https://github.com/kedacore/external-scaler-azure-cosmos-db

**Comment by zroubalik** [comment]
> > Looks good - my main comment is I don't think we necessarily need to require all scalars have a maintainer - but scalars that don't need to be clearly identified and users should consume at their own risk (maintainers won't promise to fix bugs or anything else, and will be marked as 'help needed')
> 
> Something like "Community maintained"? What would be the stimulation for anyone to become the maintainer then? Won't that end up with all scalers be community maintained?

Agree with @jeffhollan 

I doubt we are able to engage enough people to cover majority of existing scalers. What would be the result them? To drop those scalers from the next release?

**Comment by tomkerkhove** [comment]
I'm OK with adding that, but how do we convince people to take ownership of a scaler then @zroubalik @jeffhollan?

File: SCALERS.md (hunk 6 above)
Comment (tomkerkhove) [reply SCALERS.md:49]: In theory yes, I wrote this going forward but maybe it's time to evaluate if we should move that out into a Microsoft-owned organization.

I think the HTTP add-on is a better example.

However "quality" might be the wrong term here but more around release timelines, quality bars, etc and that we have the mandate to "enforce" things (ie end "responsibility/ownership")

File: SCALERS.md (hunk 5 above)
Comment (tomkerkhove) [reply SCALERS.md:37]: Then we should come up with a version for scalers as discussed in the main repo along with a deprecation policy, but that's fine for me!

Comment (tomkerkhove) [reply SCALERS.md:37]:
```suggestion
```

File: SCALERS.md (hunk 6 above)
Comment (zroubalik) [reply SCALERS.md:48]: I am okay with keeping those external scalers in KEDA repo, but would rephrase the sentence I have highlighted. Core maintainers shouldn't be responsible for that. Maintainers of that repo should be responsible, if there are no maintainers, then it should be community best effort.

**Comment by zroubalik** [comment]
> I'm OK with adding that, but how do we convince people to take ownership of a scaler then @zroubalik @jeffhollan?

SWAG and fame? 😄

File: SCALERS.md (hunk 6 above)
Comment (tomkerkhove) [reply SCALERS.md]: Updated, let me know what you think

**Review by zroubalik — APPROVED** [review]
LGTM

**Review by jeffhollan — APPROVED** [review]
Thank you so much @tomkerkhove - this looks great
